File: proyecto/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include "ast.h"
#include <string>
#include <sstream>
using namespace std;

Parser::Parser(istream* in){
  scan = new Scanner(in);
}

Parser::~Parser(){
  try{
    delete scan;
  }catch(...){}
}

AST* Parser::parse(){
  return Prog();
}

AST* Parser::Prog(){
  AST* result = Expr();
  Token* t = scan->getToken();

  if(t->getType() != eof){
    throw ParseError;
  }
  return result;
}

AST* Parser::Expr(){
  return RestExpr(Term());
}

AST* Parser::RestExpr(AST* e){
  Token* t = scan->getToken();

  if(t->getType() == add){
    return RestExpr(new AddNode(e,Term()));
  }

  if(t->getType() == sub){
    return RestExpr(new SubNode(e,Term()));
  }
  scan->putBackToken();

  return e;
}

AST* Parser::Term(){
  
  return RestTerm(Storable());
}

AST* Parser::RestTerm(AST* e){
  Token* t = scan->getToken();
  
  if(t->getType() == times){
    return RestTerm(new TimesNode(e,Storable()));
  }

  if(t->getType() == divide){
    return RestTerm(new DivideNode(e,Storable()));
  }

  if(t->getType() == modulo){
    return RestTerm(new ModuloNode(e,Storable()));
  }
  
  scan->putBackToken();

  return e;
}

AST* Parser::Storable(){

  return MemOperation(Factor());
}

AST* Parser::MemOperation(AST* result){

  Token *t = scan->getToken();

  if(t->getType() == keyword){

      if(t->getLex() == "S"){
    
      return new StoreNode(result);
      
    }else if(t->getLex() == "P"){

      return new PlusNode(result);

    }else if(t->getLex() == "M"){

      return new MinusNode(result);

    }else{
      throw ParseError;
    }
  }
  scan ->putBackToken();
  return result;
}

AST* Parser::Factor(){
  Token* t = scan->getToken();

  //cout << t->getType() << endl;
  
  if(t->getType() == number){
    istringstream in(t->getLex());
    int numero;
    in >> numero;
    return new NumNode(numero);
    
  }else if(t->getType() == identifier){
    istringstream in(t->getLex());
    string identifier;
    in >> identifier;
    return Assignable(identifier);
  }
  
  if(t->getType() == keyword){

    if(t->getLex() == "R"){

      return new RecallNode();
    }else if(t->getLex() == "C"){

      return new CleanNode();
    }
    throw ParseError;
  }

  if(t->getType() == lparen){
    AST* result = Expr();
    t = scan->getToken();
    if(t->getType() == rparen ){      
      return result;
    }
    throw ParseError;
  }
  throw ParseError;
}

AST* Parser::Assignable(string identifier){

  return  (Assign(identifier));
}

AST* Parser::Assign(string nombre){  
  Token *t = scan -> getToken();  
  NumNode* result = new NumNode(0);

  int valor;  

  if(t -> getType() == igual){
    
    result = new NumNode(Expr() -> evaluate());
    result -> assignate(nombre,result -> evaluate());
    
    
  }else{
    
    valor = result -> encontrar(nombre);    
    result  =  new NumNode(valor);    
  }
  scan -> putBackToken();  
  return result;
}

File: proyecto/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  
  def initialize(istream)
    @scan = Scanner.new(istream)
  end
  
  def parse()
    return Prog()
  end
    
  private
  def Prog()
    result = Expr()
    t = @scan.getToken()
    
    if t.type != :eof then
      raise ParseError.new
    end
    
    return result
  end
    
  def Expr() 
    return RestExpr(Term())
  end
     
  def RestExpr(e) 
    t = @scan.getToken()
    
    if t.type == :add then
         return RestExpr(AddNode.new(e,Term()))
    end
    
    if t.type == :sub then
      return RestExpr(SubNode.new(e,Term()))
    end
      
    @scan.putBackToken()
    
    return e
  end
  
  def Term()
    RestTerm(Storable())
  end
  
  def RestTerm(e)
    t = @scan.getToken()
    if t.type == :times then
      return RestTerm(TimesNode.new(e,Storable()))
    end

    if t.type == :modulo then
      return RestTerm(ModuloNode.new(e,Storable()))
    end
    
    if t.type == :divide then
      return RestTerm(DivideNode.new(e, Storable()))
    end    
    @scan.putBackToken
    return e
  end
  
  def Storable()
    return MemOperation(Factor())
  end
    
  def MemOperation(result)
    t = @scan.getToken  #obtener un token

    if t.type == :keyword then
      if t.lex == "S" then
        return StoreNode.new(result)
      elsif t.lex == "M" then
        return MinusNode.new(result)
      elsif t.lex == "P" then
        return PlusNode.new(result)
      end
      raise ParseError.new
    end
    @scan.putBackToken()  #Devuelve el caracter ingresado
      return result      
  end
       
  def Factor() 
    t = @scan.getToken    
    if t.type == :number then
      return NumNode.new(t.lex.to_i)
    end
    
    if t.type == :identifier then
      return Assignable(t.lex)
    end

    if t.type == :keyword then
      if t.lex == ?R then
        return RecallNode.new
      elsif t.lex == ?C then
        return CleanNode.new
      end
      raise ParseError.new
    end

    if t.type == :lparen then
      result = Expr()
      t = @scan.getToken
      if t.type == :rparen then
        return result
      end
      raise ParseError.new
    end
    raise ParseError.new
  end
  
  def Assignable(identifier)
    return Assign(identifier)
  end
  
  def Assign(nombre)
    t = @scan.getToken  #obtener un token
    result = NumNode.new(0)
    
    if t.type == :igual then
      result = NumNode.new(Expr().evaluate())
      result.assignate(nombre,result.evaluate())
    else
      valor = result.encontrar(nombre);
      result = NumNode.new(valor);
    end
    @scan.putBackToken
    return result    
  end  
end
